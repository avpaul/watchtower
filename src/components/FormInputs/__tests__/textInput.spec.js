import React from 'react';
import { shallow, mount } from 'enzyme';
import waitForTest from 'wait-for-expect';
import cases from 'jest-in-case';

import TextInput from '../textInput';

describe('TextInput ', () => {
  const defaultProps = {
    name: 'Name',
    parent: {
      setState: jest.fn(),
      state: {}
    },
    label: 'Name',
    length: 30,
    className: 'wt-text-input',
    type: 'text',
    placeholder: 'Name Surname',
    comment: 'No comment',
    testInput: jest.fn()
  };

  /**
   * Sets up a enzyme wrapper
   * @param object PropsOverride A object of values used to override the props passed into the component being tested
   * @param boolean shouldMount A bool value to select which type of wrapper to create
   * @return object { wrapper, props }
   */
  const setup = (propsOverride, shouldMount = false) => {
    const newProps = { ...{ ...defaultProps, ...propsOverride } };
    const wrapper = shouldMount
      ? mount(<TextInput {...newProps} />)
      : shallow(<TextInput {...newProps} />);
    return { wrapper, props: newProps };
  };

  /**
   * Tests the text input according to different scenarios
   * @param object testCase Contains values for a specified scenario
   * @param function testInputSpy A mock function to test an input check by using a function
   */
  const testTextInput = async (testCase, testInputSpy = null) => {
    const { wrapper } = setup({
      ...defaultProps,
      testInput: testCase.testInput
    });
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: testCase.value } });
    await waitForTest(() => {
      if (testInputSpy) expect(testInputSpy).toHaveBeenCalled();
      expect(wrapper.state('status')).toBe(testCase.status);
      expect(wrapper.instance().isValid()).toBe(testCase.status === 6);
    });
  };

  /**
   * Tests the set status method according to different scenarios
   * @param object testCase Contains the values for a specified scenario
   */
  const testSetStatusMethod = testCase => {
    const { wrapper } = setup();
    wrapper.instance().setStatus(...testCase.params);
    expect(wrapper.state('status')).toBe(testCase.statusNo);
  };

  it('renders as expected', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with minimal config props', async () => {
    const { wrapper } = setup({
      length: null,
      label: null,
      placeholder: null,
      className: null,
      type: null,
      comment: null,
      testInput: null
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders textarea input', () => {
    const { wrapper } = setup({ type: 'textarea' });
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the onChange action when function test is valid', async () => {
    const testInputSpy = jest.fn();
    const name = 'John Doe';
    const testInput = input => {
      testInputSpy();
      return name === input;
    };
    await testTextInput({ testInput, value: name, status: 6 }, testInputSpy);
  });

  it('focus on input successfully', () => {
    const { wrapper } = setup(null, true);
    const spyOnInputFocus = jest.spyOn(
      wrapper.instance().textInputRef.current,
      'focus'
    );
    wrapper.instance().focus();
    expect(spyOnInputFocus).toHaveBeenCalled();
  });

  cases(
    'tests the temporary state',
    async testCase => {
      const { wrapper } = setup(null, true);
      wrapper.instance().setStatus(testCase.name);
      expect(wrapper.state('status')).toBe(testCase.testStatus);
      await waitForTest(() => {
        expect(wrapper.state('status')).toBe(0);
      });
    },
    [
      {
        name: 'success',
        testStatus: 2
      },
      {
        name: 'fail',
        testStatus: 1
      }
    ]
  );

  it('sets the state of the input as valid', async () =>
    testSetStatusMethod({ params: ['valid'], statusNo: 6 }));

  it('sets the state of the input as invalid', async () =>
    testSetStatusMethod({ params: ['invalid', 'Invalid input'], statusNo: 5 }));

  it('sets the state of the input with invalid status', async () =>
    testSetStatusMethod({ params: ['correct'], statusNo: 0 }));
});
