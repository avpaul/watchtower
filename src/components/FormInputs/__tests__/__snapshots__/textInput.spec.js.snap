// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput  renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextInput
    alertText=""
    className="wt-text-input"
    comment="No comment"
    componentStateKey="inputs"
    defaultStatus={0}
    inputValue=""
    label="Name"
    length={30}
    name="Name"
    parent={
        Object {
            "setState": [Function],
            "state": Object {},
          }
    }
    placeholder="Name Surname"
    testInput={[Function]}
    type="text"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="Name"
>
          Name
</label>,
        <input
          id="Name"
          onChange={[Function]}
          placeholder="Name Surname"
          type="text"
          value=""
/>,
        <div
          className="alert--disabled"
          role="alert"
/>,
        <div
          className="wt-text-input__comment--active"
>
          No comment
</div>,
      ],
      "className": "input-group wt-text-input wt-text-input--normal mb-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Name",
          "htmlFor": "Name",
        },
        "ref": null,
        "rendered": "Name",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "Name",
          "onChange": [Function],
          "placeholder": "Name Surname",
          "type": "text",
          "value": "",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "alert--disabled",
          "role": "alert",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No comment",
          "className": "wt-text-input__comment--active",
        },
        "ref": null,
        "rendered": "No comment",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="Name"
>
            Name
</label>,
          <input
            id="Name"
            onChange={[Function]}
            placeholder="Name Surname"
            type="text"
            value=""
/>,
          <div
            className="alert--disabled"
            role="alert"
/>,
          <div
            className="wt-text-input__comment--active"
>
            No comment
</div>,
        ],
        "className": "input-group wt-text-input wt-text-input--normal mb-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name",
            "htmlFor": "Name",
          },
          "ref": null,
          "rendered": "Name",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "Name",
            "onChange": [Function],
            "placeholder": "Name Surname",
            "type": "text",
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "alert--disabled",
            "role": "alert",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No comment",
            "className": "wt-text-input__comment--active",
          },
          "ref": null,
          "rendered": "No comment",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`TextInput  renders textarea input 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextInput
    alertText=""
    className="wt-text-input"
    comment="No comment"
    componentStateKey="inputs"
    defaultStatus={0}
    inputValue=""
    label="Name"
    length={30}
    name="Name"
    parent={
        Object {
            "setState": [Function],
            "state": Object {},
          }
    }
    placeholder="Name Surname"
    testInput={[Function]}
    type="textarea"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="Name"
>
          Name
</label>,
        <textarea
          id="Name"
          onChange={[Function]}
          placeholder="Name Surname"
          required={true}
          rows="4"
          type="text"
          value=""
/>,
        <div
          className="alert--disabled"
          role="alert"
/>,
        <div
          className="wt-text-input__comment--active"
>
          No comment
</div>,
      ],
      "className": "input-group wt-text-input wt-text-input--normal mb-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Name",
          "htmlFor": "Name",
        },
        "ref": null,
        "rendered": "Name",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "Name",
          "onChange": [Function],
          "placeholder": "Name Surname",
          "required": true,
          "rows": "4",
          "type": "text",
          "value": "",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "alert--disabled",
          "role": "alert",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No comment",
          "className": "wt-text-input__comment--active",
        },
        "ref": null,
        "rendered": "No comment",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="Name"
>
            Name
</label>,
          <textarea
            id="Name"
            onChange={[Function]}
            placeholder="Name Surname"
            required={true}
            rows="4"
            type="text"
            value=""
/>,
          <div
            className="alert--disabled"
            role="alert"
/>,
          <div
            className="wt-text-input__comment--active"
>
            No comment
</div>,
        ],
        "className": "input-group wt-text-input wt-text-input--normal mb-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name",
            "htmlFor": "Name",
          },
          "ref": null,
          "rendered": "Name",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "Name",
            "onChange": [Function],
            "placeholder": "Name Surname",
            "required": true,
            "rows": "4",
            "type": "text",
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "alert--disabled",
            "role": "alert",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "No comment",
            "className": "wt-text-input__comment--active",
          },
          "ref": null,
          "rendered": "No comment",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`TextInput  renders with minimal config props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextInput
    alertText=""
    className={null}
    comment={null}
    componentStateKey="inputs"
    defaultStatus={0}
    inputValue=""
    label={null}
    length={null}
    name="Name"
    parent={
        Object {
            "setState": [Function],
            "state": Object {},
          }
    }
    placeholder={null}
    testInput={null}
    type={null}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          htmlFor="Name"
/>,
        <input
          id="Name"
          onChange={[Function]}
          placeholder={null}
          type="text"
          value=""
/>,
        <div
          className="alert--disabled"
          role="alert"
/>,
        <div
          className="null__comment--active"
/>,
      ],
      "className": "input-group null null--normal mb-3",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "htmlFor": "Name",
        },
        "ref": null,
        "rendered": null,
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "Name",
          "onChange": [Function],
          "placeholder": null,
          "type": "text",
          "value": "",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "alert--disabled",
          "role": "alert",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "className": "null__comment--active",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="Name"
/>,
          <input
            id="Name"
            onChange={[Function]}
            placeholder={null}
            type="text"
            value=""
/>,
          <div
            className="alert--disabled"
            role="alert"
/>,
          <div
            className="null__comment--active"
/>,
        ],
        "className": "input-group null null--normal mb-3",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "htmlFor": "Name",
          },
          "ref": null,
          "rendered": null,
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "Name",
            "onChange": [Function],
            "placeholder": null,
            "type": "text",
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "alert--disabled",
            "role": "alert",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "null__comment--active",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
